drivers: clock_control: Add support for nRF54H20 clock controllers

anangl opened the pull request:
Add custom clock_control API for nRF platforms that allows requesting clocks with specified minimal required attributes (accuracy, precision, and frequency). Provide an implementation of this API for FLL16M, HFXO, HSFLL, and LFCLK controllers in the nRF54H20 SoC.

anangl committed:
dts: nordic: nrf54h20: Add references to clock controllers

... so that users can obtain those for particular peripherals
with DT macros like `DT_CLOCKS_CTLR()`.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: clock_control: Add support for nRF54H20 clock controllers

Add custom clock_control API for nRF platforms that allows requesting
clocks with specified minimal required attributes (accuracy, precision,
and frequency). Provide an implementation of this API for FLL16M, HFXO,
HSFLL, and LFCLK controllers in the nRF54H20 SoC.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
soc: nordic: nrf54h: Move HSFLL trimming to clock_control driver

This trimming was done in the SoC initialization function because
there was no clock_control driver for nRF54H Series so far.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
samples: boards: nrf: Add clock_control sample

Add a sample that presents how the custom clock_control API introduced
for nRF54H20 is supposed to be used.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

hubertmis commented on drivers/clock_control/clock_control_nrf2_common.c:
```diff
@@ -0,0 +1,149 @@
+{
+	atomic_val_t prev_flags = atomic_or(&cfg->flags, FLAG_UPDATE_NEEDED);
+
+	if (prev_flags & (FLAG_UPDATE_NEEDED | FLAG_UPDATE_IN_PROGRESS)) {
```
I'm confused here. It look like line 27 results in `prev_flags` equal to `FLAG_UPDATE_NEEDED` or `0`. So what is the purpose of checking `FLAG_UPDATE_IN_PROGRESS` here?

anangl replied:
> I'm confused here. It look like line 27 results in `prev_flags` equal to `FLAG_UPDATE_NEEDED` or `0`.

No, `prev_flags` will have `FLAG_UPDATE_IN_PROGRESS` set if the update work is currently executed, and it will have `FLAG_UPDATE_NEEDED` if an update is already scheduled. In both these cases, the update work is not to be submitted (in the former case, it will be submitted at the end of `clock_config_update_end()`).

hubertmis replied:
I see. Too many reviews in a single day :)

hubertmis commented on drivers/clock_control/clock_control_nrf2_fll16m.c:
```diff
@@ -0,0 +1,201 @@
+	uint32_t active_options;
+
+	active_options = clock_config_update_begin(work);
+	if (active_options & BIT(2)) {
```
These magic bits in `active_options` are hard to follow in the review

anangl replied:
Yes, this will be reworked.

hubertmis commented on drivers/clock_control/clock_control_nrf2_hfxo.c:
```diff
@@ -0,0 +1,184 @@
+BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1,
+	     "multiple instances not supported");
+
+#define HFXO_START_TIME_US 900
```
This value should be read from BICR. It's not constant. It depends on the installed xtal

hubertmis commented on drivers/clock_control/clock_control_nrf2_hfxo.c:
```diff
@@ -0,0 +1,184 @@
+struct dev_data_hfxo {
+	struct onoff_manager mgr;
+	onoff_notify_fn notify;
+	struct k_timer timer;
```
Please note this timer is needed for FP1.0 revision. For FP1.1 it should not be necessary as a hardware event could be used instead.

hubertmis commented on drivers/clock_control/clock_control_nrf2_hsfll.c:
```diff
@@ -0,0 +1,211 @@
+	const struct device *dev = DEVICE_DT_INST_GET(0);
+	struct hsfll_dev_data *dev_data = dev->data;
+
+	k_timer_stop(&dev_data->timer);
```
What if this function gets called after the timer fires? Is it ok to call `clock_config_update_end()` multiple times for one `_begin()`?

anangl replied:
`clock_config_update_end()` checks at its beginning if an update is in progress and returns if not. But indeed, it did not clear `FLAG_UPDATE_IN_PROGRESS` at that point, so another entry could cause issues in further processing. This has been fixed.

hubertmis replied:
Could you add API docs for the common module describing design principles like this? It would be easier to follow the programmers intention while reading the code.

anangl replied:
Added.

hubertmis commented on drivers/clock_control/clock_control_nrf2_hsfll.c:
```diff
@@ -0,0 +1,211 @@
+
+static void freq_setting_applied_cb(enum dvfs_frequency_setting new_setting)
+{
+	ARG_UNUSED(new_setting);
```
This part looks invalid. `freq_setting_applied_cb()` can be called unsolicited, when other CPUs are scaling their frequencies and voltages. This function in this CPU should configure HSFLL to apply `new_setting`, even if it is higher than the last requested setting.

It can be fixed in another PR. But at least a comment like `TODO` would be good.

hubertmis replied:
And I'm missing the actual code modifying the frequency. Is it somewhere within this file?

anangl replied:
> And I'm missing the actual code modifying the frequency. Is it somewhere within this file?

It's handled in [ld_dvfs](https://github.com/zephyrproject-rtos/zephyr/blob/5d87cd654fbe2d430bcdaf39c666683abe65f290/modules/hal_nordic/nrfs/dvfs/ld_dvfs.c#L256).

anangl replied:
> This part looks invalid. `freq_setting_applied_cb()` can be called unsolicited, when other CPUs are scaling their frequencies and voltages.

Yep. I'm reworking this.

hubertmis commented on drivers/clock_control/clock_control_nrf2_hsfll.c:
```diff
@@ -0,0 +1,211 @@
+
+	k_timer_init(&dev_data->timer, hsfll_update_timeout_handler, NULL);
+
+	dvfs_service_handler_register_freq_setting_applied_callback(
+		freq_setting_applied_cb);
```
I expected to find requesting the initial frequency (the lowest one?) somewhere around here. Is it in some other place I missed?

Rafal-Nordic replied:
At least Secure Domain should set the lowest oppoint during initialisation. Without this, dvfs scaling on App core will not work.

anangl replied:
Enabling `CONFIG_NRFS_LOCAL_DOMAIN_DVFS_SCALE_DOWN_AFTER_INIT` should do the job: https://github.com/zephyrproject-rtos/zephyr/blob/5d87cd654fbe2d430bcdaf39c666683abe65f290/modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c#L291-L294
I'll add this as default.

hubertmis replied:
Now I see a race condition: the `ld_dvfs_something` module sets the initial frequency when ABB gets applied (max). In the meantime, this module is expected to set the initial frequency (probably min, or user-defined). These operations can happen in any order resulting in unknown initial frequency.

In the current design with `ld_dvfs_handler` I have a feeling that the initial frequency should be requested from clock_control to ld_dvfs_handler. The `ld_dvfs_handler` should apply the requested initial frequency immediately (if ABB has already been applied) or after ABB gets applied.

hubertmis replied:
Looks good. I've just spotted a potential race condition in ld_dvfs_handler.c:

The requested frequency is correctly set by the procedure you linked. However it might get incorrectly overwritten by `static void dvfs_service_handler_set_initial_hsfll_config(void)`. Is my observation correct?

anangl replied:
From what I can tell, `dvfs_service_handler_task()` waits for `dvfs_service_sync_sem` before calling `dvfs_service_handler_change_freq_setting()`, so `dvfs_service_handler_set_initial_hsfll_config()` needs to complete first. Unless `NRFS_DVFS_EVT_INIT_DONE` is received twice (what would mean that something is really off), it should be fine.

hubertmis replied:
Right. I didn't realize that the only invocations of these functions are protected from races with a semaphore. It's not the cleanest code (adding more calls to these functions can cause race conditions), but now it seems fine.

lstnl replied:
CONFIG_NRFS_LOCAL_DOMAIN_DVFS_SCALE_DOWN_AFTER_INIT this config was introduced to enable option to change DVFS settings from application core without other existing solution to do that. I don't know if this shouldn't be explicitly called by secdom for example when all initialization is done?

hubertmis commented on drivers/clock_control/clock_control_nrf2_hsfll.c:
```diff
@@ -167,6 +179,28 @@ static int api_get_rate_hsfll(const struct device *dev,
 {
+	const struct hsfll_dev_config *dev_config = dev->config;
+	nrf_hsfll_trim_t trim = {
+		.vsup   = sys_read32(dev_config->trim_vsup),
+		.coarse = sys_read32(dev_config->trim_coarse),
+		.fine   = sys_read32(dev_config->trim_fine),
```
It looks like the `TCOEF` trim is missing from the list.

anangl replied:
Yes, it's missing. The nrfx HAL needs to be extended in this regard as currently the `nrf_hsfll_trim_t` structure does not contain the related field.

hubertmis replied:
Do you plan to fix this in this PR?

anangl commented:
Rebased.

hubertmis commented on drivers/clock_control/clock_control_nrf2_hfxo.c:
```diff
@@ -0,0 +1,196 @@
+	} else
+#endif
+	{
+		dev_data->start_up_time = K_USEC(1000);
```
BICR misconfiguration should be treated as a fault instead of setting a default value. This hardcoded default could be different than similar defaults in other libraries (e.g. radio protocols).

lstnl commented on modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c:
```diff
@@ -182,7 +213,13 @@ static void nrfs_dvfs_evt_handler(nrfs_dvfs_evt_t const *p_evt, void *context)
-		LOG_DBG("DVFS handler EVT_OPPOINT_REQ_CONFIRMED");
+		dvfs_service_handler_clear_state_bit(DVFS_SERV_HDL_FREQ_CHANGE_REQ_PENDING_BIT_POS);
+		LOG_DBG("DVFS handler EVT_OPPOINT_REQ_CONFIRMED %d", (uint32_t)p_evt->freq);
+		if (dvfs_service_handler_get_requested_oppoint() == p_evt->freq) {
```
This is case where requested frequency is same as previous one. If it is not then it will be handled later.

anangl commented:
Rebased.

hubertmis commented on drivers/clock_control/clock_control_nrf2_lfclk.c:
```diff
@@ -0,0 +1,223 @@
+	nrfs_err_t err;
+
+	active_options = clock_config_update_begin(work);
+	while (option_idx) {
```
> can you use CLZ here to get highest bit that is set?

Isn't the optimizer smart enough? We shouldn't obscure the code with optimizations if not necessary

hubertmis replied:
There is room for improvements in GCC (and clang apparently) ;)

anangl committed:
dts: nordic: nrf54h20: Add nodes representing HFXO and LFXO oscillators

... so that their properties, like accuracy or start-up time, can be
specified for boards in devicetree.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
boards: nrf54h20dk: Add parameters of mounted HFXO and LFXO crystals

These parameters will be used by the clock controller drivers.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

nordic-krch commented on drivers/clock_control/clock_control_nrf2_hfxo.c:
```diff
@@ -0,0 +1,196 @@
+					NRF_LRCCONF_EVENT_HFXOSTARTED));
+
+	if (dev_data->notify) {
+		dev_data->notify(&dev_data->mgr, 0);
```
i see in the comment that it may not be started.

nordic-krch commented on modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c:
```diff
@@ -152,11 +165,29 @@ static void dvfs_service_handler_set_initial_hsfll_config(void)
+
+static void dvfs_service_handler_scaling_finish_delay_timeout(struct k_timer *timer)
+{
+	if (timer) {
```
it was part of #73783 and no longer in that pr.

nordic-krch commented on modules/hal_nordic/nrfs/dvfs/Kconfig:
```diff
@@ -26,6 +26,11 @@ config NRFS_LOCAL_DOMAIN_DOWNSCALE_SAFETY_TIMEOUT_US
+config NRFS_LOCAL_DOMAIN_DOWNSCALE_FINISH_DELAY_TIMEOUT_US
+	int "Additional delay to let secdom finish dowscale procedure in us"
+	range 1 10000000
+	default 1000
```
it was part of #73783 and no longer in that pr.

nordic-krch commented on modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c:
```diff
@@ -152,11 +165,29 @@ static void dvfs_service_handler_set_initial_hsfll_config(void)
+{
+	if (timer) {
+		dvfs_service_handler_scaling_finish(
+			*(enum dvfs_frequency_setting *)timer->user_data);
```
it was part of #73783 and no longer in that pr.

nordic-krch commented on modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c:
```diff
@@ -196,15 +233,20 @@ static void nrfs_dvfs_evt_handler(nrfs_dvfs_evt_t const *p_evt, void *context)
+			static enum dvfs_frequency_setting freq;
+
+			freq = p_evt->freq;
+			dvfs_service_scaling_finish_delay_timer.user_data = (void *)&freq;
```
it was part of #73783 and no longer in that pr.

hubertmis commented on drivers/clock_control/clock_control_nrf2_common.c:
```diff
@@ -0,0 +1,158 @@
+
+	active_options = cfg->flags_snapshot & BIT_MASK(ONOFF_CNT_MAX);
+
+	return active_options ? (uint8_t)(31 - __builtin_clz(active_options))
```
I would move `(uint8_t)(31 - __builtin_clz(active_options))` to a `static inline` function with a name indicating what it is intended to calculate. Right now I'm guessing what was your intention :)

bjarki-andreasen replied:
Done in new PR

hubertmis reviewed (commented):
The missing TCOEF in HSFLL trimming is the last thing I would like to agree on before I approve. It's fine for me to implement it out of this PR, but we need at least to track it to make sure we don't forget

hubertmis commented on drivers/clock_control/clock_control_nrf2_fll16m.c:
```diff
@@ -0,0 +1,239 @@
+
+static void activate_fll16m_mode(struct fll16m_dev_data *dev_data, uint8_t mode)
+{
+	/* TODO: change to nrf_lrcconf_* function when such is available. */
+	NRF_LRCCONF010->CLKCTRL[0].SRC = mode;
```
While debugging a clock issue with @KAGA164 I realized we've forgotten to manage LRCCONF's power. In some scenarios, it results in losing clock mode requests.

Setting `mode` to any other value than the hardware default (open-loop) should be preceded by `NRF_LRCCONF010->POWERON.MAIN = 1`. Setting `mode` to the hardware default should be followed by resetting `POWERON.MAIN`.

Please note, that other components could require setting POWERON.ACT or POWER.MAIN independently. One example is HFXO clock_control instance. So accessing LRCCONF010->POWERON.MAIN requires a multi-client helper (another on-off instance?)

bjarki-andreasen replied:
Added in new PR

hubertmis commented on drivers/clock_control/clock_control_nrf2_hfxo.c:
```diff
@@ -0,0 +1,192 @@
+
+	dev_data->notify = notify;
+
+	nrf_lrcconf_event_clear(NRF_LRCCONF010, NRF_LRCCONF_EVENT_HFXOSTARTED);
```
Like described in FLL16M, this must be preceded by setting `NRF_LRCCONF010->POWERON.MAIN = 1`.

bjarki-andreasen replied:
Added in new PR

hubertmis commented on drivers/clock_control/clock_control_nrf2_hfxo.c:
```diff
@@ -0,0 +1,192 @@
+
+static void onoff_stop_hfxo(struct onoff_manager *mgr, onoff_notify_fn notify)
+{
+	nrf_lrcconf_task_trigger(NRF_LRCCONF010, NRF_LRCCONF_TASK_STOPREQHFXO);
```
And this must be followed with `NRF_LRCCONF010->POWERON.MAIN = 0` if no other user requires keeping the main power domain on. The "If no other user..." part requires users counting, like another on-off instance?

bjarki-andreasen replied:
Added in new PR

hubertmis reviewed (changes requested):
I've just realized there is overlooked bug in this implementation.

bjarki-andreasen commented:
This PR will be continued from the new PR #77365

bjarki-andreasen commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -723,6 +742,7 @@
 				reg = <0x99c000 0x1000>;
 				status = "disabled";
 				cc-num = <16>;
+				clocks = <&lfclk>;
```
Added in new PR
